### PR TITLE
Document covariance vs invariance

### LINKED ISSUE #2034: Document covariance vs invariance
Questions about covariance and invariance of collection types such as `List[...]` and `Dict[...]` are pretty common, but we have no documentation about them. It would be nice to have good documentation, since they are some of the trickier mypy concepts to understand, especially together with how type inference works in mypy.

Potential things to do:
- Add section about this and give an overview of how things work and how to define covariant generic types. Also mention contravariance. Give some examples.
- Discuss how `List` and `Dict` are invariant (also others) and why, with a motivating example. Mention that `Iterable`, `Sequence` and others are covariant, and using them can work around some issues.
- Discuss how type inference works for things like `[A(), B()]` if `B` is a subclass of `A`.
- Provide some workaround in the section on common issues. For example, giving an explicit annotation can help. Another thing that may help is generating a copy of a collection with a different type, such as `list_a = list(list_b)  # type: List[A]`. Casts and `# type: ignore` can also be used, but they aren't safe. Covariant types such as `Sequence` can also sometimes be used to work around issues.
- Also have some examples that use union types, as this often often comes up together with union types.

### PR DESCRIPTION
**Pull request by ilevkivskyi**
Fixes #2034

### CODE REVIEW

File: docs/source/generics.rst [hunk 1]
```diff
@@ -219,6 +219,84 @@ or a deserialization method returns the actual type of self. Therefore
...
+type variables defined with special keyword arguments ``covariant`` or
+``contravariant``. For example:
+
+.. code-block:: python
+
+   from typing import Generic, TypeVar
+   T_co = TypeVar('T_co', covariant=True)
+
+   class Box(Generic[T_co]):  # this type is declared covariant
+       def __init__(self, content: T_co) -> None:
```
Comment (ilevkivskyi) [inline docs/source/generics.rst]: mypy complains at this line (covariant type in contravariant position), but I believe this is OK, since ``__init__`` is special. (otherwise, how would I put something into a covariant container?)

File: docs/source/common_issues.rst [hunk 2]
```diff
@@ -180,6 +180,50 @@ not support ``sort()``) as a list and sort it in-place:
         # Type of x is List[int] here.
         x.sort()  # Okay!
 
+Invariance vs covariance
+------------------------
+
+Most mutable generic collections are invariant, as well all user defined
```
Comment (JelleZijlstra) [inline docs/source/common_issues.rst]: This is a run-on sentence. Maybe reword as "Most mutable generic collections are invariant, and mypy considers all user-defined generic classes invariant by default (see :ref:`variance-of-generics` for motivation)."

File: docs/source/generics.rst [hunk 3]
```diff
@@ -219,6 +219,84 @@ or a deserialization method returns the actual type of self. Therefore
...
+* ``Callable`` is an example of type that behaves contravariant in types of
+  arguments, namely ``Callable[[Employee], int]`` is a subtype of
+  ``Callable[[Manager], int]``. To understand this, consider a function:
+
+  .. code-block:: python
+
+     def salaries(staff: List[Manager],
+                  accountant: Callable[[Manager], int]) -> List[int]: ...
+
+  this function needs a callable that can calculate a salary for mangers, and
```
Comment (JelleZijlstra) [inline docs/source/generics.rst]: managers

mangers don't get a salary :)

File: docs/source/generics.rst [hunk 4]
```diff
@@ -219,6 +219,84 @@ or a deserialization method returns the actual type of self. Therefore
...
+  ``T`` if ``MyCovGen[B, ...]`` is always a subtype of ``MyCovGen[A, ...]``.
+* A generic class ``MyContraGen[T, ...]`` is called contravariant in type
+  variable ``T`` if ``MyContraGen[A, ...]`` is always a subtype of
+  ``MyContraGen[B, ...]``.
+* A generic class ``MyInvGen[T, ...]`` is called invariant in ``T`` if neither
+  of the above is true.
+
+Let us illustrate this by few simple examples:
+
+* ``Union`` is an example of type that is covariant in all variables, namely
```
Comment (JelleZijlstra) [inline docs/source/generics.rst]: of *a* type. Better yet, shorten this to "Union is covariant in all variables: ``Union[Cat, int]`` is a subtype ..."

File: docs/source/generics.rst [hunk 5]
```diff
@@ -219,6 +219,84 @@ or a deserialization method returns the actual type of self. Therefore
...
+
+  .. code-block:: python
+
+     def salaries(staff: List[Manager],
+                  accountant: Callable[[Manager], int]) -> List[int]: ...
+
+  this function needs a callable that can calculate a salary for mangers, and
+  if we give it a callable that can calculate a salary for an arbitrary
+  employee, then it is still safe.
+* ``List`` is an example of invariant generic type. Naively, one would think
```
Comment (JelleZijlstra) [inline docs/source/generics.rst]: remove "example of"

File: docs/source/generics.rst [hunk 6]
```diff
@@ -219,6 +219,84 @@ or a deserialization method returns the actual type of self. Therefore
...
+         things.append(Shape())
+
+     my_things: List[Circle] = []
+     add_one(my_things)     # This may appear safe, but...
+     my_things[0].rotate()  # ...this will fail
+
+  Another example of invariant type is ``Dict``, most mutable containers
+  are invariant.
+
+By default, mypy assumes that all user defined generics are invariant,
```
Comment (JelleZijlstra) [inline docs/source/generics.rst]: This is a run-on sentence, so you should change the comma into a period. Also, "user-defined" is usually written as one word I think.

**Comment by ilevkivskyi** [comment]
@JelleZijlstra Thank you for review! I made corresponding changes in a new commit.

File: docs/source/generics.rst (hunk 1 above)
Comment (JukkaL) [inline docs/source/generics.rst]: Yes, as discussed elsewhere, mypy should allow this.

Comment (ilevkivskyi) [inline docs/source/generics.rst]: I was thinking about this some time and it looks like we could make a simple exclusion for ``__init__``, I will make a PR now.

**Comment by JukkaL** [comment]
Thanks for the doc updates!
